pkg/apis/v1alpha1/nodeoverlay.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+type NodeOverlayList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []NodeOverlay `json:"items"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=nodeoverlays,scope=Cluster,categories=karpenter
+type NodeOverlay struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +required
+	Spec NodeOverlaySpec `json:"spec"`
+}
+
+type NodeOverlaySpec struct {
+	// Selector matches against simulated nodes and modifies their scheduling properties. Matches all if empty.
+	// +optional
+	Selector []v1.NodeSelectorRequirement `json:"selector,omitempty"`
+	// PricePercent modifies the price of the simulated node (PriceAdjustment + (Price * PricePercent / 100)).
```

Comment:
The more complex the pricing constructs are, the more complex our algorithm gets -- at some point, we can't solve everything. We also need to start somewhere and make progress. One option would be to try to write some automation that removes the NodeOverlay when the savings plan is exhausted.

Reply:
> I don't know how you could detect this.

It will query the Compute Savings Plan and calculate how much it costs for the current existing nodes. If the total hourly cost is over the commitment, it should be the original on-demand price. It may not be very accurate, but it could solve part of the problem.

> At some point, we can't solve everything.

Absolutely correct. I am thinking about whether we can write a component to handle this kind of situation. Karpenter will query the real price from the component service.

This is what I am thinking.

```yaml
# ec2 sp node overlay
---
kind: NodeOverlay
metadata:
  name: discount
spec:
  selector:
    matchLabels:
      karpenter.sh/capacity-type: on-demand
  priceAPIEndpoint: http://price.karpenter.svc:8080
```

```yaml
# ec2 sp node pool
apiVersion: karpenter.sh/v1beta1
kind: NodePool
metadata:
  name: default
spec:
  template:
    spec:
      requirements:
        - key: kubernetes.io/arch
          operator: In
          values: ["amd64"]
        - key: kubernetes.io/os
          operator: In
          values: ["linux"]
        - key: karpenter.k8s.aws/instance-category
          operator: In
          values: ["t"]
        - key: karpenter.k8s.aws/instance-generation
          operator: Gt
          values: ["3"]
      nodeClassRef:
        apiVersion: karpenter.k8s.aws/v1beta1
        kind: EC2NodeClass
        name: default
  limits:
    limitsAPIEndpoint: http://limits.karpenter.svc:8080
  weight: 100 # highest
---
# other node pool
```

So, using the EC2 SP node pool, it will query the limits and price from another component, and the complex logic will be calculated in that component. When the EC2 SP is finished, we will use another node pool.